### PR TITLE
geo/wkt: implement parser for points with Z and M dimensions

### PR DESCRIPTION
**Pull request by andyyang890**
This patch adds a parser that is capable of parsing WKT
representations of points with Z and M dimensions.

Refs: #53091

Release note: None

### CODE REVIEW

File: pkg/geo/wkt/lex.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+func isNumRune(r rune) bool {
+	switch r {
+	case '-', '.':
+		return true
+	default:
+		return unicode.IsDigit(r)
+	}
+}
+
+// Lex a number.
```
Comment (otan) [inline pkg/geo/wkt/lex.go:139]: nit: `num lexes a number` (have func name at front)

File: pkg/geo/wkt/lex.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+		if !isNumRune(c) {
+			break
+		}
+		b.WriteRune(l.next())
+	}
+
+	fl, err := strconv.ParseFloat(b.String(), 64)
+	if err != nil {
+		l.lastErr = &LexError{
+			problem: "bad number",
```
Comment (otan) [inline pkg/geo/wkt/lex.go:155]: nit: invalid number

File: pkg/geo/wkt/lex.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		if unicode.ToUpper(l.peek()) == 'M' {
+			l.next()
+			b.WriteRune('M')
+		}
+	}
+
+	ret := getKeywordToken(b.String())
+	if ret == eof {
+		l.lastErr = &LexError{
+			problem: "bad keyword",
```
Comment (otan) [inline pkg/geo/wkt/lex.go:122]: nit: invalid keyword

File: pkg/geo/wkt/wkt.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import "github.com/twpayne/go-geom"
+
+// Unmarshal accepts a string and parses it to a geom.T.
+func Unmarshal(wkt string) (geom.T, error) {
+	wktlex := &wktLex{line: wkt}
+	ret := wktParse(wktlex)
+	if wktlex.lastErr != nil {
+		return nil, wktlex.lastErr
+	} else if ret != 0 {
```
Comment (otan) [inline pkg/geo/wkt/wkt.go:23]: nit: omit `else` if we returned early. the linter will complain too.

File: pkg/geo/wkt/lex_test.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+			desc:           "2D point with extra comma",
+			input:          "POINT(0, 0)",
+			expectedErrStr: `parse error: could not parse "POINT(0, 0)"`,
+		},
+	}
+
+	for _, tc := range errorTestCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			got, err := Unmarshal(tc.input)
+			require.Nil(t, got)
```
Comment (otan) [inline pkg/geo/wkt/lex_test.go:111]: nit: usually for errors we don't assert the values other than error.

**Review by otan — COMMENTED** [review]
very nice! some small nits

**Review by otan — APPROVED** [review]
i'm accepting this to unblock you in the morning, and there are only nits!
i trust you to fix them up :)

File: pkg/geo/wkt/lex_test.go (hunk 5 above)
Comment (andyyang890) [reply pkg/geo/wkt/lex_test.go:111]: Oh what's the reasoning behind that? Is it because the returned value is assumed to be garbage if there is an error?

Comment (otan) [reply pkg/geo/wkt/lex_test.go]: usually yeah the contract is that it can be garbage.
(seen cases where this does not apply, but it's rare)

**Comment by andyyang890** [comment]
bors r=otan

**Pull request merged**